### ISSUE TITLE
useSearch unsafe parameter decoding

### ISSUE THREAD

**Issue by slybridges** [molefrog/wouter#400]
First of all, thanks for Wouter v3!

`useSearch()` hook seems to aggressively decode parameters, which can lead to unsafe parameter objects being created when used in combination with `URLSearchParams`

**Expectations** : using `useSearch()` on a page with query parameters `?search=foo%26parameter_injection%3Dbar` and parsing it with the code below should yield:

```javascript
{ search: "foo&parameter_injection=bar" }
```

**Actuals**: `useSearch()` returns `search=foo&parameter_injection=bar` which, when used with the code below gives you:

```javascript
{ search: "foo", parameter_injection: "bar" }
```

The following code can repro the issue:

```javascript
const searchString = useSearch()
const searchParams = new URLSearchParams(searchString)
const query = {}
for (const [key, value] of searchParams.entries()) {
  query[key] = value
}
console.log(query)
```

This can lead to user input adding or overriding query parameters in the app.

**Comment by molefrog** [comment]
Thank you for spotting this. Could you check if replacing `decodeURIComponent` with `decodeURI` when decoding a query string solves the issue?

**Comment by slybridges** [comment]
Hi @molefrog,
Yes, `decodeURI` does preserve query parameters as-is and would prevent the user from injecting parameters.

```js
decodeURIComponent("?search=foo%26parameter_injection%3Dbar")
# '?search=foo&parameter_injection=bar' -> BAD

decodeURI("?search=foo%26parameter_injection%3Dbar")
# '?search=foo%26parameter_injection%3Dbar' -> GOOD
```

**Comment by molefrog** [comment]
Fixed in v3.0.0-rc.3
